lib/db_utils.go

```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
+}
+
+func DBGetAccessGroupIdsForMember(db *badger.DB, snap *Snapshot,
+	accessGroupMemberPublicKey PublicKey) (_accessGroupIdsMember []*AccessGroupId, _err error) {
+
+	var ret []*AccessGroupId
+	var err error
+	err = db.View(func(txn *badger.Txn) error {
+		ret, err = DBGetAccessGroupIdsForMemberWithTxn(txn, snap, accessGroupMemberPublicKey)
+		return err
+	})
+	if err != nil && err != badger.ErrKeyNotFound {
+		return nil, errors.Wrapf(err, "DBGetAccessGroupIdsForMember: Problem getting access group ids for member")
+	}
+	return ret, nil
+}
+
+func DBGetAccessGroupIdsForMemberWithTxn(txn *badger.Txn, snap *Snapshot,
+	accessGroupMemberPublicKey PublicKey) (_accessGroupIdsMember []*AccessGroupId, _err error) {
+
+	prefix := _dbSeekPrefixForAccessGroupMemberEntry(accessGroupMemberPublicKey)
+	keysFound := _enumerateKeysOnlyForPrefixWithTxn(txn, prefix)
+
+	// Decode found keys.
+	accessGroupIds := []*AccessGroupId{}
+	for _, keys := range keysFound {
+		accessGroupMemberPublicKeyFromKey, accessGroupOwnerPublicKey, accessGroupKeyName, err := _dbDecodeKeyForAccessGroupMemberEntry(keys)
+		if err != nil {
+			return nil, errors.Wrapf(err, "DBGetAccessGroupIdsForMemberWithTxn: Problem decoding access group id")
+		}
+		if !bytes.Equal(accessGroupMemberPublicKey.ToBytes(), accessGroupMemberPublicKeyFromKey.ToBytes()) {
+			return nil, fmt.Errorf("DBGetAccessGroupIdsForMemberWithTxn: "+
+				"Access group member public key from key (%v) does not match expected access group member public key (%v)",
+				accessGroupMemberPublicKey, accessGroupMemberPublicKeyFromKey)
+		}
+
+		accessGroupId := NewAccessGroupId(&accessGroupOwnerPublicKey, accessGroupKeyName.ToBytes())
+		accessGroupIds = append(accessGroupIds, accessGroupId)
+	}
+	return accessGroupIds, nil
+}
+
+func DBPutAccessGroupMemberEntry(db *badger.DB, snap *Snapshot, blockHeight uint64,
+	accessGroupMemberEntry *AccessGroupMemberEntry, accessGroupOwnerPublicKey PublicKey, groupKeyName GroupKeyName) error {
+	var err error
+	err = db.Update(func(txn *badger.Txn) error {
+		return DBPutAccessGroupMemberEntryWithTxn(txn, snap, blockHeight,
+			accessGroupMemberEntry, accessGroupOwnerPublicKey, groupKeyName)
+	})
+	if err != nil {
+		return errors.Wrapf(err, "DBPutAccessGroupMemberEntry: Problem putting access group member entry")
+	}
+	return nil
+}
+
+func DBPutAccessGroupMemberEntryWithTxn(txn *badger.Txn, snap *Snapshot, blockHeight uint64,
+	accessGroupMemberEntry *AccessGroupMemberEntry, accessGroupOwnerPublicKey PublicKey, groupKeyName GroupKeyName) error {
+
+	if accessGroupMemberEntry == nil || accessGroupMemberEntry.AccessGroupMemberPublicKey == nil {
+		return fmt.Errorf("DBPutAccessGroupMemberEntryWithTxn: accessGroupMemberEntry is nil or " +
+			"accessGroupMemberEntry.AccessGroupMemberPublicKey is nil")
+	}
+	if reflect.DeepEqual(groupKeyName.ToBytes(), BaseGroupKeyName().ToBytes()) {
+		glog.Errorf("DBPutAccessGroupMemberEntryWithTxn: groupKeyName is empty")
```

Comment:
what do we want to do here? do we save base key? if so, I would remove the if statement. reflect.DeepEqual is expensive relatively.